KEYCLOAK-16633 Prevent deletion of internal clients.

artur-baltabayev opened the pull request:
Implements functionality that prevents the deletion of internal clients, as described in the Jira issue.

artur-baltabayev committed:
KEYCLOAK-16633 Prevent deletion of internal clients.

artur-baltabayev commented:
https://issues.redhat.com/browse/KEYCLOAK-16633

abstractj commented:
Closing this PR based as this issue was already fixed. See: https://issues.redhat.com/browse/KEYCLOAK-16633

vramik commented on services/src/main/java/org/keycloak/services/managers/ClientManager.java:
```diff
@@ -93,7 +94,8 @@ public static ClientModel createClient(KeycloakSession session, RealmModel realm
 
     public boolean removeClient(RealmModel realm, ClientModel client) {
-        if (realm.removeClient(client.getId())) {
+        if (!isInternalClient(realmManager.getSession().realms().getRealmsStream(), client.getClientId())
+                && realm.removeClient(client.getId())) {
```
There should be check that the realm is actually master realm.

artur-baltabayev replied:
Fixed. The "isInternalClient()" method now instantly returns false if it's the master realm.

vramik commented on services/src/main/java/org/keycloak/services/managers/ClientManager.java:
```diff
@@ -355,4 +357,7 @@ private boolean showClientCredentialsAdapterConfig(ClientModel client) {
         return authenticator.getAdapterConfiguration(client);
     }
 
+    private boolean isInternalClient(Stream<RealmModel> stream, String clientId) {
+        return stream.anyMatch(realm -> clientId.equals(realm.getName() + "-realm"));
+    }
```
There is no need to go through all existing realms here, with large number of realms it could become costly operation. Can it be used `session.realms().getRealmByName()` instead?

artur-baltabayev replied:
Fixed. The realm name is now parsed from the client (if a "-realm" suffix is present) and thus only a single query is made.

vramik commented on services/src/main/java/org/keycloak/services/resources/admin/ClientResource.java:
```diff
@@ -221,8 +222,12 @@ public void deleteClient() {
+            adminEvent.operation(OperationType.DELETE).resourcePath(session.getContext().getUri()).success();
+        }
+        else {
+            throw new ForbiddenException();
```
I think we shouldn't throw the `ForbiddenException` here. Please take a look at `ErrorResponseException` instead.

artur-baltabayev replied:
Could you maybe explain why it would be better to throw an ErrorResponseException here ? A ForbiddenException seems fitting for me in this particular case.

artur-baltabayev replied:
I changed it to be a ErrorResponseException. I'm assuming the reason for it is that you can't be entirely sure that the reason for failure is an internal client. Therefore i included a "Bad Request" response into the ErrorResponseException. This seems to be consistent with the other ones.

vramik replied:
That's right, also in UI you get nice error message instead of "Unexpected error"

vramik reviewed (changes requested):
Thank you @artur-baltabayev for the PR. I have few comments, please see inline. 

In addition to clients from master realm with the suffix -realm, there are other internal clients that should be prevented from being deleted, e.g. account, etc. (the full list might be discussed in mailing list)

abstractj commented:
@artur-baltabayev would you like to follow up on @vramik's comments?

artur-baltabayev commented:
@abstractj @vramik Hello, sorry i have been very busy the last days with another contribution. I will look at the comments tomorrow and integrate them. Thank you very much for the feedback !!!

artur-baltabayev commented:
@vramik @abstractj As for the other internal clients, there is a "defaultClients" static list, which i think is the right choice here because it might be dangerous to keep a non-centralized list as it might not be updated in the future.

What do you think about this ?

vramik commented:
> @vramik @abstractj As for the other internal clients, there is already a "defaultClients" static list, which i think is the right choice here because it might be dangerous to keep a non-centralized list as it might not be updated in the future.
> 
> What do you think about this ?

+1 for `Constants.defaultClients` static list

artur-baltabayev commented:
@vramik @abstractj I integrated your feedback into the PR. I also changed some additional stuff to improve the overall code quality :
1. Removed a .* import.
2. Removed a unused method.

I wanted to say that i'm not 100% sure on the ErrorResponseException. The other exception cases seem to include a INVALID_REQUEST error and BAD_REQUEST response code, so i thought i'd make it consistent and include it also in this case. Since it probably shouldn't be treated as a point of failure for internal clients only it seemed fitting.

As always, any feedback is greatly appreciated !!

vramik commented on services/src/main/java/org/keycloak/services/managers/ClientManager.java:
```diff
@@ -366,4 +367,19 @@ private boolean showClientCredentialsAdapterConfig(ClientModel client) {
         return authenticator.getAdapterConfiguration(client);
     }
 
+    private boolean isInternalClient(String realmName, String clientId) {
+        if (!"master".equals(realmName)) {
+            return false;
+        }
+
+        final String internalClientSuffix = "-realm";
+
+        if (clientId.endsWith(internalClientSuffix)) {
+            return realmManager.getSession().realms()
+                    .getRealmByName(
+                            clientId.substring(0, clientId.length() - internalClientSuffix.length())) != null;
+        }
+
+        return defaultClients.contains(clientId);
+    }
```
```suggestion
    private boolean isInternalClient(String realmName, String clientId) {
        if (defaultClients.contains(clientId) return true;
        
        if (!"master".equals(realmName)) {
            return false;
        }

        final String internalClientSuffix = "-realm";

        if (clientId.endsWith(internalClientSuffix)) {
            return realmManager.getSession().realms()
                    .getRealmByName(
                            clientId.substring(0, clientId.length() - internalClientSuffix.length())) != null;
        }

        return false;
    }
```

artur-baltabayev replied:
Fixed. I also improved the testing, so it now tests the removal on every default client.

vramik reviewed (changes requested):
Thank you @artur-baltabayev for the update. The PR looks good. There is one remaining thing from my POV. 

Due to fact that clients with `-realm` suffix are internal in case they belong to `master` realm, but all other internal clients (those from `Constrants.defaultClients`) are not just in `master` realm, I've suggested a change, please see inline. 

I'll take a closer look at ErrorResponseException, but from first look it looks fine to me. 

Thanks again for all the work.

vramik reviewed (approved):
Thank you @artur-baltabayev.

@abstractj from my POV it is ready to be merged.

abstractj commented:
Tested it locally and LGTM. Thank you, @artur-baltabayev! And thanks @vramik, for reviewing.
